rust/src/operations/restore.rs

```diff
@@ -0,0 +1,288 @@
+    let state_to_restore_files_set =
+        HashSet::<Add>::from_iter(state_to_restore_files.iter().cloned());
+    let latest_state_files_set = HashSet::<Add>::from_iter(latest_state_files.iter().cloned());
+
+    let files_to_add: Vec<Add> = state_to_restore_files
+        .into_iter()
+        .filter(|a: &Add| !latest_state_files_set.contains(a))
+        .map(|mut a: Add| -> Add {
+            a.data_change = true;
+            a
+        })
+        .collect();
+
+    let deletion_timestamp = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis() as i64;
+    let files_to_remove: Vec<Remove> = latest_state_files
+        .into_iter()
+        .filter(|a: &Add| !state_to_restore_files_set.contains(a))
+        .map(|a: Add| -> Remove {
+            Remove {
+                path: a.path.clone(),
+                deletion_timestamp: Some(deletion_timestamp),
+                data_change: true,
+                extended_file_metadata: Some(false),
+                partition_values: Some(a.partition_values.clone()),
+                size: Some(a.size),
+                tags: a.tags,
+            }
+        })
+        .collect();
+
+    if !ignore_missing_files {
+        check_files_available(object_store.as_ref(), &files_to_add).await?;
+    }
+
+    let metrics = RestoreMetrics {
+        num_removed_file: files_to_remove.len(),
+        num_restored_file: files_to_add.len(),
+    };
+
+    let mut actions = vec![];
+    let protocol = if protocol_downgrade_allowed {
+        action::Protocol {
+            min_reader_version: table.get_min_reader_version(),
+            min_writer_version: table.get_min_writer_version(),
+        }
+    } else {
+        action::Protocol {
+            min_reader_version: max(
+                table.get_min_reader_version(),
+                snapshot.min_reader_version(),
+            ),
+            min_writer_version: max(
+                table.get_min_writer_version(),
+                snapshot.min_writer_version(),
+            ),
+        }
+    };
+    actions.push(Action::protocol(protocol));
+    actions.extend(files_to_add.into_iter().map(Action::add));
+    actions.extend(files_to_remove.into_iter().map(Action::remove));
+
+    let commit = prepare_commit(
+        object_store.as_ref(),
+        &DeltaOperation::Restore {
+            version: version_to_restore,
+            datetime: datetime_to_restore.map(|time| -> i64 { time.timestamp_millis() }),
+        },
+        &actions,
+        None,
+    )
+    .await?;
+    let commit_version = snapshot.version() + 1;
+    match try_commit_transaction(object_store.as_ref(), &commit, commit_version).await {
+        Ok(_) => {}
+        Err(err @ TransactionError::VersionAlreadyExists(_)) => {
+            return Err(err.into());
+        }
+        Err(err) => {
+            object_store.delete(&commit).await?;
+            return Err(err.into());
+        }
+    }
```

Comment:
Just making sure I understand. For restore we always want to fail, if any other operation was performed on the table?

Reply:
Yes, if any other operation was performed, the restore is invalid because the ADD and REMOVE files are computed based on latest snapshot state and state of restored version.